### ISSUE TITLE
Add the option for Safari 12 Webdriver to use '--legacy' argument

### ISSUE THREAD

**Issue by charlesmorelle84** [SeleniumHQ/selenium#6431]
OS:  OSX High Sierra
Selenium Version:  3.14.0
Browser:  Safari 12

The implementation of the Safari 12  WebDriver now supports the W3C endpoints **by default**, which may be causing issues if you're using a Framework like Watir, which aren't compatible yet: https://developer.apple.com/documentation/webkit/macos_webdriver_commands_for_safari_12_and_later

Browser Version:  Version 12.0 (13606.2.11)

**Comment by charlesmorelle84** [comment]
> Usage: safaridriver [options]
	--legacy                  Use the legacy Selenium WebDriver protocol.

**Comment by gzfrancisco** [comment]
Just found this problem on my CI after upgrading the Safari version. Does anyone know how to hot-fixit  it?

**Comment by charlesmorelle84** [comment]
I was able to temporarily fix it by updating this file /Users/[user_name]/.rvm/gems/ruby-X.X.X/gems/selenium-webdriver-X.X.X/lib/selenium/webdriver/safari/service.rb

And adding this line:
![screen shot 2018-09-19 at 17 06 00](https://user-images.githubusercontent.com/22692532/45788288-7018ef80-bc2e-11e8-861a-8024388b0494.png)

**Comment by lmtierney** [comment]
How is Watir not compatible? It should be, please raise an issue if it is not. Selenium should also be compatible, so are you having issues with it as well?

**Comment by lmtierney** [comment]
Closing this as you can already add the `--legacy` flag, see:

```
driver = Selenium::WebDriver.for(:safari, driver_opts: {args: '--legacy'})
```

Or Watir:
```
browser = Watir::Browser.new(:safari, driver_opts: {args: '--legacy'})
```
If you have other issues with Selenium and Safari 12, please open a new issue with the details.

**Comment by charlesmorelle84** [comment]
Thank you @lmtierney .

It's working by declaring it with the args value:
`$driver = Watir::Browser.new(:safari, driver_opts: {args: '--legacy'})`

**Comment by gzfrancisco** [comment]
I confirm, the solution works with the Selenium WebDriver. Thanks.

**Comment by mjdease** [comment]
@lmtierney how can this flag be set using the node selenium-webdriver package (v3.6)?

**Comment by lmtierney** [comment]
Using the service https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/safari_exports_ServiceBuilder.html and the `addArguments` method.

For further questions, please head over to the [selenium user group](https://groups.google.com/forum/#!forum/selenium-users).

**Comment by mjdease** [comment]
For reference here's how I got it working:
```js
new webdriver.Builder()
    .usingServer(await new safari.ServiceBuilder().addArguments('--legacy').build().start())
    .forBrowser('safari')
    .build();
```

The builder API doesn't seem to have a `setSafariService` method like there are for other browsers (eg `setChromeService`) so I needed to start the service directly and set it as the server to use.

**Comment by mjdease** [comment]
@tonyaveniros That's the new JS language feature [async/await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) which is supported in recent versions of node (7.6+ according to MDN). 
The [DriverService start method](https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/remote/index_exports_DriverService.html#start) returns a promise that resolves with the server url to use.

Using the promises directly, the code would be like:
```js
new safari.ServiceBuilder().addArguments('--legacy').build().start().then((serverUrl) => {
    return new webdriver.Builder()
        .usingServer(serverUrl)
        .forBrowser('safari')
        .build();
});
```

**Comment by tonyaveniros** [comment]
Thanks for that. I'm still a bit turned around on this.

Upgraded to node v8.12.0 and now I'm getting the error message
```
.usingServer(await new safari.ServiceBuilder().addArguments('--legacy').build().start())
                 ^^^^^

SyntaxError: missing ) after argument list
```

Using Promises directly does not seem to work either as the next function returns
```
	driver.wait(function () {
		       ^
TypeError: Cannot read property 'wait' of undefined

```
I'm wondering if your use case is testing remotely using **selenium grid**: mine is for testing **locally**...

**Comment by tonyaveniros** [comment]
Got it working with:
```
new webdriver.Builder()
    .usingServer( new safari.ServiceBuilder().addArguments('--legacy').build().start())
    .forBrowser('safari')
    .build();
```
using node v8.12.0

**Comment by Tenzian** [comment]
Is there a way to pass this argument to Selenium Standalone Server?

**Comment by AlexSdet** [comment]
**Issue still happening for me when using '--legacy' flag**

Safari: 12.0.3
Ruby: 2.3.3
Selenium: 3.4.0
Using 'Capybara' framework with selenium webdriver.

Add argument like suggested above:
![still_issue](https://user-images.githubusercontent.com/48072475/53529040-52cf3380-3aa0-11e9-9df6-4edc134a3375.png)

But still seeing an issue:
**Request body does not contain required parameter 'capabilities'.**
  /Users/username/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/selenium-webdriver-3.4.0/lib/selenium/webdriver/remote/response.rb:69

Any ideas what im doing wrong ?

Thanks

**Comment by cgoldberg** [comment]
already answered in #6026  ... see: https://github.com/SeleniumHQ/selenium/issues/6026#issuecomment-468070211

please don't post the same comment in multiple issues.

**Comment by TheSpyder** [comment]
It looks like Safari 12.1 (which just shipped with macOS 10.14.4) silently removed support for the `--legacy` argument. The nodejs `selenium-webdriver` package hasn't been updated since 3.6.0 in October 2017 (except for a 4.0-alpha release in Jan 2018) so I can no longer use selenium to test Safari.

**Comment by philipwalton** [comment]
@TheSpyder just noticed the same. I'm also now unable to get my tests running for Safari 12.1. I'd love to hear if you find a solution.

**Comment by TheSpyder** [comment]
@philipwalton until there are signs that selenium wants to update their nodejs bindings, we're working on a rewrite to https://webdriver.io

**Comment by shs96c** [comment]
The JS bindings marked as 4.0.0-alpha-1 are stable. The maintainer marked them this way as we've not yet shipped Selenium 4.0 itself. We'll correct this and ship the updated files to npm.
